### ISSUE TITLE
Worrisome trends in ELF self-play games

### ISSUE THREAD

**Issue by Crazyeight101** [gcp/leela-zero#1541]
I've been looking a lot at the ELF self-play games the past several weeks, and I noticed two very peculiar trends, especially regarding games that ended with a low number of moves (generally <150):
1. The game ends when 1 player has mistakenly tenuki'd from a large liberty race that it would otherwise win and the sente is clearly not big enough to justify the tenuki, or it considers a liberty race to be lost when it is very clearly won and resigns early.
http://zero.sjeng.org/view/612de0a4574358723ffc134208974be0b21b7349abb06b4c428a546e05062f6c?viewer=wgo
http://zero.sjeng.org/view/0c4a651c64813e31097799c23fd226d8b1c6c946682344b2c2d08a87170edb30?viewer=wgo
2. The game reaches a position that's highly favorable for 1 player, either through an early capture or a bad sequence, at which point both players begin to play an arbitrary number of apparently random moves before ending the game.
http://zero.sjeng.org/view/cbb693fdf00eefa0131bdea1326143099c10f901f07f99e0d67ace735b38bb20?viewer=wgo
http://zero.sjeng.org/view/8635da122734c72b61008316e4e24c804c7bef58cc60c4a6b9d7762b00fa18da?viewer=wgo

I've provided a handful of examples here, but I've witnessed these happen dozens and dozens of times.  This seems to be very commonly occurring.  Also, comparatively, in the LZ network self-play games, I've observed the first behavior considerably less often than in the ELF self-play games, and I haven't observed the second behavior at all.

This made me curious about a couple things:
Could this be related to training a 15 block network with weights from a 20 block network?
What impact is this having on training the LZ network?

What this seems to be suggesting to me is that, in the case of the liberty races, LZ is being made aware of its weakness when it comes to large-scale liberty races, in the sense that Elf weights that were trained to handle them are leading LZ to positions where they are created, but LZ still lacks the ability to correctly evaluate them.  In games where ELF weights don't lead to large liberty races, which tend to be the games that ended with a higher number of moves (generally 150-300), I think ELF is showing LZ better ways to play positions that can actually be correctly evaluated on a 15 block network.

As for the extremely random bad moves that I see occur at the end of games over and over again, I have no idea what impact that could potentially have, but intuition tells me it's related to some underlying issue with how ELF is trained and its very sharp value head.

Wouldn't this seem to make using ELF in self-play games on a smaller network a double-edged sword?  While LZ is clearly benefiting from the ELF self-play with quick short-term growth, I'm concerned it might also be having unintended consequences in the long-term by introducing biases in the network with a large number of false results that probably can't be uncovered until LZ moves to a same size or larger network than ELF was trained on.  Worst case, in my opinion, would be that this causes LZ to get stuck in some local extrema later on.  This is all conjecture on my part, but the trends in the self-play games are definitely real.

Thoughts?

**Comment by Splee99** [comment]
I'm curious too, but I couldn't find obvious random move. Could you explain in more details?

**Comment by WYVYW** [comment]
They specifically did T=1 randomness and noise… in order to try to get away a bit from ELF, and yes, it may mean slower progress on the pure strength side. That's a tradeoff.

https://github.com/gcp/leela-zero/issues/1403

**Comment by Crazyeight101** [comment]
@Splee99 as an example, starting at move 49, look at http://zero.sjeng.org/view/cbb693fdf00eefa0131bdea1326143099c10f901f07f99e0d67ace735b38bb20?viewer=wgo

You can go look through the shorter ELF self-play games and find copious examples of this.

@WYVYW granted I haven't looked at the code to see how the randomness is actually implemented, what I'm describing is probably not what you're talking about.  Up to a certain point in the game, when the position is still undecided, the engine will play it in a reasonable manner, with perhaps only 2-4 moves that are questionable or unusual and clearly the result of introducing noise into the training.  Once a favorite becomes clear, say through a fairly large early capture, then the engine will inexplicably start playing unusual and bad moves, for both players, until it decides to end the game.  This is something I haven't seen in LZ network self-play games.

Also, while the string of random moves at the end of that kind of game is an interesting trend, I feel that the other one I pointed out is more meaningful, as the engine is almost reliably throwing won games.

**Comment by Ishinoshita** [comment]
@Crazyeight101 
> Once a favorite becomes clear, say through a fairly large early capture, then the engine will inexplicably start playing unusual and bad moves, for both players, until it decides to end the game. This is something I haven't seen in LZ network self-play games.

I have observed this behaviour on plenty of no-resign games. After typically 300 moves, one side is clearly behind and starts playing strange moves (from human standard), i.e. moves that look random or, worse, bad moves like failing to save on large group that could easily have lived or filling its own eyes up to putting itself in atari.
This clearly results from the search been lost when dealing with very adverse values. The search makes all moves look equal, since the game is lost anyway, while the policy alone would probably play 'better'. Nothing illogical, nothing shocking but for human eyes. However, one difference vs what you are reporting is that I noted that only the losing side tends to play weirdly, the winning side calmly captures most of the opponent groups, one by one, with his help.
I'm wondering what is the impact of such portion of these kind of selfplay games for training (such games can and often do last up to +500, thus contribute to +200-300 training samples subject to this kind of behaviour).

**Comment by l1t1** [comment]
@Crazyeight101 I saw the white move 50, it is very silly. but black move 51 also very strange, why it not move D7?
![cbb6](https://user-images.githubusercontent.com/23301704/41200395-6fc01832-6c93-11e8-977d-01f6e0bb174a.png)
![cbb62](https://user-images.githubusercontent.com/23301704/41200407-8b68d70e-6c93-11e8-829c-58f6b3120c1c.png)

**Comment by l1t1** [comment]
@Crazyeight101 I saw the white move 50, it is very silly. but black move 51 also very strange, why it not move D7?
![cbb6](https://user-images.githubusercontent.com/23301704/41200395-6fc01832-6c93-11e8-977d-01f6e0bb174a.png)
![cbb62](https://user-images.githubusercontent.com/23301704/41200407-8b68d70e-6c93-11e8-829c-58f6b3120c1c.png)

**Comment by l1t1** [comment]
can you use https://github.com/pnprog/goreviewpartner/releases and elf weight to analyze the game above? @Crazyeight101

**Comment by Splee99** [comment]
I don't think you can reproduce the noise in the training game.

**Comment by Marcin1960** [comment]
Will ELF dominate LZ at 250K games?

**Comment by pondturtle** [comment]
For what it's worth I have seen LZ playing some really stupid kyu worth blunders at lower visit number. Seems pretty typical behaviour. Usually it needed 60 000+ visits to actually play correct move in those situations.

**Comment by jkiliani** [comment]
@Ttl investigated the optimum visit count to use for training games, in #1416, and those results rather clearly indicate that the optimum is at a lower visit number than what we're using right now. If anything, visits should be lowered, not raised.

**Comment by pondturtle** [comment]
Fair enough. Doing a lot of analysing my games with Lizzie I just noticed LZ suggesting some stupid blunders even I could see. on rare occasion before it accumulated a lot of visits. Often in situations like the one that is talked about here - big semeais.

**Comment by Marcin1960** [comment]
@Crazyeight101 "Could this be related to training a 15 block network with weights from a 20 block network?"

Would switching to 20x192 net help? Without going to the 20x256 big and slow net?

What would be speed penalty?

**Comment by l1t1** [comment]
lz148 will be the longest live15x192 weight soon


no | Upload Date | Hash | Size | Elo | Games | Training no
-- | -- | -- | -- | -- | -- | --
148 | 2018-06-08 04:31 | d0187996 | 192x15 | 11455 | 90388 | 7792166
125 | 2018-04-18 04:43 | e8601c38 | 192x15 | 10511 | 90912 | 6853079
